src/authorization/WebAclReader.ts

```diff
@@ -147,122 +129,115 @@ export class WebAclReader extends PermissionReader {
   }
 
   /**
-   * Sets the correct values for non-acl permissions such as create and delete.
-   * Also adds the correct values to indicate that having control permission
-   * implies having read/write/etc. on the acl resource.
+   * Finds the ACL data relevant for all the given resources.
+   * The input set will be modified in place.
+   *
+   * @param targets - Targets to find ACL data for.
    *
-   * The main reason for keeping the control value is so we can correctly set the WAC-Allow header later.
+   * @returns A map linking ACL resources to the relevant identifiers.
    */
-  private updateAclPermissions(aclPermissions: AclPermission, isAcl: boolean): AclPermission {
-    if (isAcl) {
-      return {
-        read: aclPermissions.control,
-        append: aclPermissions.control,
-        write: aclPermissions.control,
-        create: aclPermissions.control,
-        delete: aclPermissions.control,
-        control: aclPermissions.control,
-      };
+  private async getAclMatches(targets: Set<ResourceIdentifier>): Promise<Map<AclMatch, ResourceIdentifier[]>> {
+    this.logger.debug(`Searching ACL data for ${[ ...targets ].map((id): string => id.path).join(', ')}`);
+
+    const storeMap = new Map<AclMatch, ResourceIdentifier[]>();
+
+    // Loop over the targets until we found matching ACL resources for each of them
+    while (targets.size > 0) {
+      // Start with the longest identifier as the matching ACL might also be relevant for other resources
+      const longest = reduceIterable(targets, (long, identifier): ResourceIdentifier =>
+        identifier.path.length > long.path.length ? identifier : long, { path: '' });
+      const aclMatch = await this.getAclRecursive(longest);
+      const matchingIdentifiers: ResourceIdentifier[] = [];
+      storeMap.set(aclMatch, matchingIdentifiers);
+      // Find all targets that use the found ACL resource for authorization.
+      // Specifically, these are the paths that are a substring of the input path and contain the response path.
+      for (const target of targets) {
+        if (!longest.path.includes(target.path)) {
```

Comment:
I would make our assumptions VERY explicit here. And perhaps even move those assumptions out. Or maybe this questions needs to be asked to a separate oracle rather than encoding the ACL/document relationship here ourselves.

Also, does it help that much with performance? Maybe better to just pass a cache to `getAclRecursive`.

Reply:
I was writing out why this works and is correct (since we find the first resource that has an acl resource, so everything inbetween would also end up with the same result when calling `getAclRecursive), but then I just realized I could find a counterexample where this actually doesn't work.

So, an example where this is the incorrect solution: the longest identifier is `/foobar`, but one of the other parameters is `/foo`. `/foobar.acl` does not exist so it finds its permissions in `/.acl`, but `/foo.acl` does exist. With the current code `/foo` would check `/.acl`. So yes, my cheeks are totally red.

The solution would be to cut off everything after the last `/` of the longest identifier in case it is an identifier, but I do understand that you probably have less trust in my idea now 😄. Although it does help in the amount of calls that have to be done since in practice most calls will contain 2 identifiers at most: a resource and its parent container. And these will quite often have the same acl so doing it like this reduces the amount of checks we have to do in half.

Using a `CachedResourceSet` would also help indeed if we introduce the change mentioned above where only the identifiers get found first and the stores are only checked later. This still has some more checks though.